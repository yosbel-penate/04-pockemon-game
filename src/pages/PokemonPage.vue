<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>
    <div v-else>
        <h1>Quién es este pokémo?</h1>
        <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
        <PokemonOptions 
            :pokemons = "pokemonArr"
             @selection="checkAnswer"
        /> 
        <template v-if="showAnswer" class="fade-in">
            <h2>{{message}}</h2>
            <button @click="newGame">
                Nuevo juego
            </button>
        </template>
    </div>
</template>

<script>
import PokemonOptions from "@/components/PokemonOptions";
import PokemonPicture from "@/components/PokemonPicture";
import getPokemontOptions from "@/helpers/getPokemonOptions";

export default {
    name: 'PokemonPage',

    components:{ PokemonOptions, PokemonPicture },
    data(){
        return {
            pokemonArr:[],
            pokemon: null,
            showPokemon:false,
            showAnswer:false,
            message: null
        }
    },
    methods: {
        async mixPokemonArrayAndSelectOne() {
            this.pokemonArr = await getPokemontOptions()
            this.pokemon = this.selectRandomOnePokemon(this.pokemonArr)
        },
        selectRandomOnePokemon(pokemonArr = []){
            const randomIntegerNumber = this.getRandomIntegerNumber(4)
            return pokemonArr[randomIntegerNumber]
        },
        getRandomIntegerNumber(interval){
            return Math.floor(Math.random()*interval)
        },
        checkAnswer(pokemonId){
            this.showAnswer = true
            if (this.pokemon.id === pokemonId) {
                this.message = `Selecionaste correctamente a ${this.pokemon.name}.`
                this.showPokemon = true
            } else{
                this.message = `Oops, era ${this.pokemon.name}`
            }
        },
        newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr = []
            this.pokemon = null
            this.mixPokemonArrayAndSelectOne()
        }
    },
     mounted(){
        this.mixPokemonArrayAndSelectOne()
    }
};
</script>

<style scoped>

</style>