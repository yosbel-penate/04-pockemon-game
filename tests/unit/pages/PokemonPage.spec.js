import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from '@/pages/PokemonPage'
import { pokemonsMock } from "../mocks/pokemons.mock"

describe('PokemonPage component', ()=> {
    let wrapper
    beforeEach(()=>{
        wrapper = shallowMount(PokemonPage)
    })
    test('debe de match con el snapshop cuando cargan lo pokemons', () => { 
        const wrapper = shallowMount(PokemonPage,{
            data(){
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon:false,
                    showAnswer:false,
                    message: null
                }
            },
        })
        expect(wrapper.html()).toMatchSnapshot()

     })
    test('debe de llamar al metodo mixPokemonArray al montar', () => { 
        const moxPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArrayAndSelectOne')
        const wrapper = shallowMount(PokemonPage)
        expect(moxPokemonArraySpy).toHaveBeenCalled()
    })
    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => { 
        const wrapper = shallowMount(PokemonPage,{
            data(){
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon:false,
                    showAnswer:false,
                    message: null
                }
            },
        })
        const pictureComponent = wrapper.find('pokemon-picture-stub')
        const optionComponet = wrapper.find('pokemon-options-stub')

        expect(pictureComponent.exists()).toBeTruthy()
        expect(pictureComponent.attributes('pokemonid')).toBe('1')
        expect(optionComponet.exists()).toBeTruthy()

     })
    test('pruebas con checkAnswer', async () => { 
        const wrapper = shallowMount(PokemonPage,{
            data(){
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon:false,
                    showAnswer:false,
                    message: null
                }
            },
        })
        await wrapper.vm.checkAnswer(5)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showAnswer).toBeTruthy()

    })

})