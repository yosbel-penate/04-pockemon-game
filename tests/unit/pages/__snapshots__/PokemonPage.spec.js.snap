// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe de match con el snapshop cuando cargan lo pokemons 1`] = `
<div>
  <h1>Quién es este pokémo?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
